### DISCUSSION TITLE
Model doesn't learn

### DISCUSSION THREAD

**Discussion by gkrisp98** [Project-MONAI/MONAI#3700] [Ideas]
Hi, I am new to the field of medical images. I am working with 435 MRIs and I have to classify them to 5 classes. I have created a subject dataset both for nifti and DICOM files but neither of them is learning. The training and validation accuracy range from 20 to 22% which is just random.
I used the transforms RandomAffine(), RescaleIntensity(), RandomNoise and CropOrPad() to make the shape of the objects (160,224,224) as it was (160,384,384).
As of now I used a efficient net 3D, a resnet3d and a resnet18 from monai but none of them seems to learn anything.
Is there any advice for what I can do to make my models start learning ?

**Comment by Nic-Ma** [comment]
Hi @ericspod @dongyang0122 ,

I think you guys are deep learning research experts, could you please help share some advice here?

Thanks in advance.

**Comment by dongyang0122** [comment]
Hi @gkrisp98,

It could be potentially multiple reasons causing an ineffective learning process. First, you can check the step for data pre-processing (since you did not mention in your descriptions). MRI has greater appearance variation compared to CT images. And the transform "NormalizeIntensityd" is recommended to use for MRI as data pre-processing. Second, you can simply the problem to make sure the learning process is proper. For instance, you can convert the 5-class recognition problems into a 2-class one (via merging some class labels), or disable the data augmentation at the moment, to make sure that the model is learning something. Last not the least, you can further check your scripts and compare to MONAI examples (https://github.com/Project-MONAI/tutorials/tree/master/3d_classification). It is possible that some necessary lines are missing to cause such ineffective learning.

BTW, if you can share your scripts, we can take a quick look and provide some insights.

Best,
Dong

**Reply by gkrisp98** [reply]
Thanks for the response! Here is some code:

The subject dataset:
```
import torchio as tio

def load_mri(mri_path):
  root = "/content/drive/MyDrive/subjects"
  full_path = os.path.join(root, mri_path)
  subject = tio.Subject(t1 = tio.ScalarImage(full_path))
  return subject

import glob
patient = [a.split("/")[-1] for a in sorted(glob.glob(f"/content/drive/MyDrive/subjects/" + "/*"))]

ids = [a.rpartition("subjects/")[2] for a in sorted(glob.glob(f"/content/drive/MyDrive/subjects/" + "/*" + "/*"))]

def makeSubjects(root_dir = "/content/drive/MyDrive/subjects"):
  dataframe = pd.read_csv("/content/drive/MyDrive/KL_grade.csv", sep = ';')
  labels = {}
  id = list(dataframe["id"])
  grades = list(dataframe["grade"])
  for i,g in zip(id, grades):
    labels[str(i).zfill(5)] = g

  subjects_list = []

  for i, idx in enumerate(ids):
    sub = tio.Subject(
        load_mri(ids[i]),
        label = labels[patient[i]],
    )
    subjects_list.append(sub)
  return subjects_list

preprocess = tio.RescaleIntensity((-1, 1))
augmentation = tio.Compose([
    tio.RandomAffine(),
    preprocess,
    tio.RandomNoise(),
    tio.CropOrPad((160,224,224))
])

subjects_list = MakeSubjects()
subjects_dataset = tio.SubjectsDataset(subjects_list, transform=augmentation)

```

For model, training and validation:

```
import monai
from monai.networks.nets import resnet18
model.conv1 = nn.Conv3d(1, 64, kernel_size=(7, 7, 7), stride=(1, 2, 2), padding=(3, 3, 3), bias=False)
model.fc = nn.Linear(in_features=512, out_features=5, bias=True)
model.cuda()

def training(epoch, model, train_loader, optimizer, criterion):
  "Training over an epoch"
  metric_monitor = MetricMonitor()
  model.train()

  for i, traindata in enumerate(train_loader):
    images = F.interpolate(traindata['t1'][tio.DATA], scale_factor=(1,1,1)).to('cuda:0')
    labels = traindata['label'].to('cuda:0')
    output = model(images)


    loss = criterion(output, labels)
    accuracy = calculate_accuracy(output, labels)
    metric_monitor.update("Loss", loss.item())
    metric_monitor.update("Accuracy", accuracy)
    images.detach()
    labels.detach()
    
    optimizer.zero_grad()
    loss = Variable(loss, requires_grad = True)

    loss.backward()
    optimizer.step()
  print("[Epoch: {epoch:03d}] Train      | {metric_monitor}".format(epoch=epoch, metric_monitor=metric_monitor))
  return metric_monitor.metrics['Loss']['avg'], metric_monitor.metrics['Accuracy']['avg']

def validation(epoch, model, valid_loader, criterion):
  "Validation over an epoch"
  metric_monitor = MetricMonitor()
   model.eval()
  with torch.no_grad():

      for i, traindata in enumerate(train_loader):
        images = F.interpolate(traindata['t1'][tio.DATA], scale_factor=(1,1,1)).to('cuda:0')
        labels = traindata['label'].to('cuda:0')
        output = model(images)
        loss = criterion(output,labels) 
        loss = Variable(loss, requires_grad = True)
        
        accuracy = calculate_accuracy(output, labels)
        metric_monitor.update("Loss", loss.item())
        metric_monitor.update("Accuracy", accuracy)
        images.detach()
        labels.detach()
  print("[Epoch: {epoch:03d}] Validation | {metric_monitor}".format(epoch=epoch, metric_monitor=metric_monitor))
  return metric_monitor.metrics['Loss']['avg'], metric_monitor.metrics['Accuracy']['avg']
```

And the calls for training and validation (I have 2 other classes for metrics and early stopping but they are irrelevant to the issue I think so I didn't provide them):
```
import numpy as np

num_epochs = 50
use_early_stopping = True
use_scheduler = True

optimizer = torch.optim.Adam(model.parameters(), lr=1e-3, betas=(0.9, 0.999), eps=1e-08, weight_decay=1e-3)
criterion = torch.nn.CrossEntropyLoss(weight = normedWeights)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.9)

train_losses , train_accuracies = [],[]
valid_losses , valid_accuracies = [],[]

if use_early_stopping:
    early_stopping = EarlyStopping(patience=20, verbose=True, delta=1e-4)

for epoch in range(1, num_epochs+1):
    
    train_loss, train_accuracy = training(epoch,model,train_loader,optimizer,criterion)
    valid_loss, valid_accuracy = validation(epoch,model,valid_loader,criterion)
    
    if use_scheduler:
        scheduler.step()
        
    train_losses.append(train_loss)
    train_accuracies.append(train_accuracy)
    valid_losses.append(valid_loss)
    valid_accuracies.append(valid_accuracy)
          
    if use_early_stopping: 
        early_stopping(valid_loss, model)
        
        if early_stopping.early_stop:
            print('Early stopping at epoch', epoch)
            model.load_state_dict(torch.load('/content/drive/MyDrive/HOPE-DCM-EFF.pth'))
            break
  
plt.plot(range(1,len(train_losses)+1), train_losses, color='b', label = 'training loss')
plt.plot(range(1,len(valid_losses)+1), valid_losses, color='r', linestyle='dashed', label = 'validation loss')
plt.legend(), plt.ylabel('loss'), plt.xlabel('epochs'), plt.title('Loss'), plt.show()
  
plt.plot(range(1,len(train_accuracies)+1),train_accuracies, color='b', label = 'training accuracy')
plt.plot(range(1,len(valid_accuracies)+1),valid_accuracies, color='r', linestyle='dashed', label = 'validation accuracy')
plt.legend(), plt.ylabel('loss'), plt.xlabel('epochs'), plt.title('Accuracy'), plt.show()
```

**Reply by dongyang0122** [reply]
At first glance, the script looks fine to me. A quick test would be using the same data for training and validation as the sanity check to see if the model can quickly overfit the data.

**Reply by gkrisp98** [reply]
I have tested it and it does not overfit :/

**Reply by dongyang0122** [reply]
Few lines in your scripts could cause the issues.

```
images.detach()
labels.detach()
loss = Variable(loss, requires_grad = True)
```

They look strange to me. Not sure why you put them there. You can try to remove them from training and validation functions. Meanwhile, it would be great if you can share your training curves (of training/validation loss and accuracy curves).

**Reply by gkrisp98** [reply]
Ok I will run some epochs without these lines and I will post the curves. Thanks a lot for your advices!

**Reply by gkrisp98** [reply]
So, here are the curves with early stopping at 10 epochs
![acc and loss](https://user-images.githubusercontent.com/57015579/150807897-7c864974-1856-450b-be27-fbc2ef6285ab.PNG)
:

**Reply by dongyang0122** [reply]
It now appears that the training is starting to take effect. The training loss starts to drop and the accuracy rises.

**Reply by gkrisp98** [reply]
Then should I disable early stopping and just run it for 50 (or more) epochs ? But shoudn't the val loss and acc still be saturating ?

**Reply by dongyang0122** [reply]
You can try to disable early stopping and see if the model overfits the data.
